pkg/pushsync/pushsync.go

```diff
@@ -119,6 +123,15 @@ func (ps *PushSync) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream)
 		return swarm.ErrInvalidChunk
 	}
 
+	//if p's address is closer to the chunk than my address, then simply store it and return
```

Comment:
To summarize, storer node, being closest to the chunk, finds the second closest peer for replication, and sends the chunk. To prevent infinite loops, nodes that are second closest, in other words, nodes that receive chunks from peers with closer addresses, simply store the chunk and exit early out of the handler method. 

Please help me understand if I am missing any details.

Reply:
@zelig the requirement you're setting here is problematic since to determine whether a chunk is in the neighborhood of the node we're getting the chunk from presupposes that node's connectivity which I think will not hold. The other problem is that we'd need to introduce neighborhood depth as a dependency here and I was trying to avoid that as much as possible, with the hope that the pricer actually detects whether some is or not within the NN, returning the correct price then.

Reply:
> To summarize, storer node, being closest to the chunk, finds the second closest peer for replication, and sends the chunk. To prevent infinite loops, nodes that are second closest, in other words, nodes that receive chunks from peers with closer addresses, simply store the chunk and exit early out of the handler method.
> 
> Please help me understand if I am missing any details.


If a node N receives  a push sync message of chunk C from its peer P:
- and N is itself the peer closest to the chunk C,  then it sends it to all its nearest neighours (which just store the chunk and do no forwarding  or  responding) and (concurrently) responds with  a receipt to P.
- and N is  not  the  closest peer, but the sending peer is within its neighbourhood, i.e., 
     - PO(N,P)  >= Depth(N), then
          - if P is the closest peer  to  C (ie., PO(N,C) > PO(N',C) for all N' peers of P), then N just stores the chunk, does no forwarding or responding
          - if N is closer to C than P (ie., PO(N,C) > PO(P,C)) then N forwards it  to the closest peer (can assume good faith and that P is not connected to N' which IS the node closest to the chunk) 
          - if N is not closer to C than P (i.e., PO(N,C) <= PO(P,C)), then N just stores the chunk, does no forwarding or responding (assuming in good faith that P is not connected to N' which IS the node closest to the chunk) 
- if P is not in N's neighbourhood then N forwards to a node